pkg/generate/code/resource_reference.go

```diff
@@ -0,0 +1,233 @@
+					"WithReferencesResolvedCondition(&resource{ko}, "+
+					"fmt.Errorf(\"'%s' field should not be present when using"+
+					" reference field '%sRef'\"))\n", indent,field.Names.Camel,
+					field.Names.Camel)
+				out += fmt.Sprintf("%s}\n", indent)
+
+				// If the field is required, make sure either Ref or original
+				// field is present in the resource
+				if field.IsRequired() {
+					out += fmt.Sprintf("%sif ko.Spec.%sRef == nil &&"+
+						" ko.Spec.%s == nil {\n", indent, field.Names.Camel,
+						field.Names.Camel)
+					fmt.Sprintf("%s// No need to add reference resolved"+
+						" condition because reference is not present\t", indent)
+					out += fmt.Sprintf("%s\treturn &resource{ko},"+
+						" fmt.Errorf(\"At least one of '%s' or '%sRef' field"+
+						" should be present\")\n", indent, field.Names.Camel,
+						field.Names.Camel)
+					out += fmt.Sprintf("%s}\n", indent)
+				}
+
+				fIndentLevel := indentLevel
+				fIndent := indent
+				out += fmt.Sprintf("%s// Checking Referenced Field"+
+					" %sRef\n", fIndent, field.Names.Camel)
+				if field.ShapeRef.Shape.Type == "list" {
+					out += fmt.Sprintf("%sif ko.Spec.%sRef != nil &&"+
+						" len(ko.Spec.%sRef) > 0 {\n", fIndent,
+						field.Names.Camel, field.Names.Camel)
+					out += fmt.Sprintf("%s\treferencePresent = true\n",
+						fIndent)
+					out += fmt.Sprintf("%s\tresolvedReferences :="+
+						" []*string{}\n", fIndent)
+					// arrw stands for AWSResourceReferenceWrapper
+					out += fmt.Sprintf("%s\tfor _, arrw := range"+
+						" ko.Spec.%sRef {\n", fIndent, field.Names.Camel)
+					out += fmt.Sprintf("%s\t\tarr := arrw.From\n",
+						fIndent)
+					out += readReferencedResource(field, strings.Repeat("\t",
+						fIndentLevel+2), apiReaderVarName, contextVarName,
+						resVarName)
+					out += checkReferencedResource(field, strings.Repeat("\t",
+						fIndentLevel+2))
+					out += fmt.Sprintf("%s\t\tresolvedReferences ="+
+						" append(resolvedReferences, obj.%s)\n", fIndent,
+						field.FieldConfig.References.Path)
+					out += fmt.Sprintf("%s\t}\n", fIndent)
+					out += fmt.Sprintf("%s\tko.Spec.%s ="+
+						" resolvedReferences\n", fIndent, field.Names.Camel)
+					out += fmt.Sprintf("%s}\n", fIndent)
+				} else {
+					out += fmt.Sprintf("%sif ko.Spec.%sRef != nil &&"+
+						" ko.Spec.%sRef.From != nil {\n", fIndent,
+						field.Names.Camel, field.Names.Camel)
+					out += fmt.Sprintf("%s\treferencePresent = true\n",
+						fIndent)
+					out += fmt.Sprintf("%s\tarr := ko.Spec.%sRef.From\n",
+						fIndent, field.Names.Camel)
+					out += readReferencedResource(field, strings.Repeat("\t",
+						fIndentLevel+1), apiReaderVarName, contextVarName,
+						resVarName)
+					out += checkReferencedResource(field, strings.Repeat("\t",
+						fIndentLevel+1))
+					out += fmt.Sprintf("%s\tko.Spec.%s = obj.%s\n",
+						fIndent, field.Names.Camel,
+						field.FieldConfig.References.Path)
+					out += fmt.Sprintf("%s}\n", fIndent)
+				}
```

Comment:
I think it would actually be easier and cleaner to have this generator function be a *template* that produces a function that looks up a *single* resource via a resource reference wrapper field.

For example, imagine a template that looked like this:

```handlebars
{{ range $field := .CRD.Fields }}
{{ if $field.HasReference }}
// resolveReferenceFor{{ $field.Names.Camel }} resolves references for {{ .CRD.Kind }}.Spec.{{ $field.Names.Camel }}
// and returns any error that occurred attempting to resolve the reference
func (rm *resourceManager) resolveReferenceFor(
	ctx context.Context,
	apiReader client.Reader,
	ns types.Namespace,
	ref *ackv1alpha1.AWSResourceReferenceWrapper,
	res acktypes.AWSResource,
) error {
	if ref != nil && ref.From != nil {
		ko := rm.concreteResource(res)
		nsName := types.NamespacedName{Namespace: ns, Name: *ref.From.Name}
		obj := acksvcv1alpha1.{{ .CRD.Kind }}{}
		err := apiReader.Get(ctx, nsName, &obj)
		if err != nil {
			return err
		}
		if ackcond.ResourceSynced(res) {
			return ackerr.ReferencedResourceNotSynced(*ref.From.Name)
		}
		if obj.{{ $field.FieldConfig.References.Path }} == nil {
			return ackerr.ReferencedResourceNotSynced(*ref.From.Name)
		}
		ko.Spec.{{ $field.Names.Camel }} = obj.{{ $field.FieldConfig.References.Path }}
}

{{ end }}
```